### PR TITLE
move away from rxjava Observable to ArrayBlockingQueue for metadata wait...

### PR DESCRIPTION
**Pull request by stevenzwu**
... messages

### CODE REVIEW

File: suro-kafka-producer/src/main/java/com/netflix/suro/sink/kafka/KafkaSink.java [hunk 1]
```diff
@@ -340,12 +330,16 @@ public void call(Throwable throwable) {
     public void close() {
         isOpened = false;
         try {
+            executor.shutdown();
```
Comment (metacret) [inline suro-kafka-producer/src/main/java/com/netflix/suro/sink/kafka/KafkaSink.java]: metadataWaitingQueue.take() will be blocked forever and executor.awaitTermination(10, TimeUnit.SECONDS) will always take 10 seconds. So, we need to set up a timeout with poll(). The example can QueuedSink.java in suro-core. If you want to make things simpler, before calling shutdown(), you can push the 'command' message to metadataWaitingQueue. In the thread loop, if the message is 'command' message, it should skip the message and break the loop.

**Comment by stevenzwu** [comment]
got you. Iet me try again.

**Comment by stevenzwu** [comment]
pushed the poison-msg change. verified with a simple test.

```
package com.netflix.suro.sink.kafka;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.netflix.suro.message.MessageContainer;
import com.netflix.suro.message.StringMessage;
import org.junit.Test;

import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class ExecutorTest {

    private final static MessageContainer SHUTDOWN_POISON_MSG = new StringMessage("suro-KafkaSink-shutdownMsg-routingKey",
            "suro-KafkaSink-shutdownMsg-body");

    @Test
    public void shutdown() throws Exception {
        final BlockingQueue<MessageContainer> metadataWaitingQueue = new ArrayBlockingQueue<MessageContainer>(1024);
        ExecutorService executor = Executors.newSingleThreadExecutor(
                new ThreadFactoryBuilder().setDaemon(false).setNameFormat("KafkaSink-MetadataFetcher-%d").build());
        executor.submit(new Runnable() {
            @Override
            public void run() {
                while(true) {
                    final MessageContainer message;
                    try {
                        message = metadataWaitingQueue.poll(1, TimeUnit.SECONDS);
                    } catch (InterruptedException e) {
                        continue;
                    }
                    if(message == null) {
                        System.out.println("got nothing");
                        continue;
                    }
                    // check poison msg for shutdown
                    if(message == SHUTDOWN_POISON_MSG) {
                        break;
                    }
                }
            }
        });
        long start = System.currentTimeMillis();
        metadataWaitingQueue.offer(SHUTDOWN_POISON_MSG);
        executor.shutdown();
        executor.awaitTermination(10, TimeUnit.SECONDS);
        System.out.println(String.format("shutdown took %d ms", System.currentTimeMillis() - start));
    }
}
```

**Comment by stevenzwu** [comment]
do you mean add the ExecutorTest.java?

**Comment by stevenzwu** [comment]
I guess you meant close  latency. added this to TestKafkaSink.java

```
    /**
     * open will success because localhost is a resolvable hostname
     * note that there is no broker running at "localhost:1"
     */
    @Test
    public void testCloseLatency() throws Exception {
        String sinkstr = "{\n" +
                "    \"type\": \"kafka\",\n" +
                "    \"client.id\": \"kafkasink\",\n" +
                "    \"bootstrap.servers\": \"localhost:1\",\n" +
                "    \"kafka.etc\": {\n" +
                "          \"acks\": \"1\"\n" +
                "      }\n" +
                "}";
        KafkaSink sink = jsonMapper.readValue(sinkstr, new TypeReference<Sink>(){});
        sink.open();
        Assert.assertTrue(sink.isOpened());
        long start = System.currentTimeMillis();
        sink.close();
        long closeLatency = System.currentTimeMillis() - start;
        System.out.println("closeLatency = " + closeLatency);
        // close should take less than 500 ms
        assertTrue(closeLatency < 500);
    }
```

**Pull request merged**